File: src/09.tree/05.437.ts
import { TreeNode } from "./01.100";

// 双递归遍历查找
// export function pathSum(root: TreeNode | null, targetSum: number): number {
//     if (!root) {
//         return 0;
//     }

//     return (
//         dfs(root, targetSum) +
//         pathSum(root.left, targetSum) +
//         pathSum(root.right, targetSum)
//     );
// }

// function dfs(root: TreeNode | null, targetSum: number): number {
//     if (!root) {
//         return 0;
//     }
//     targetSum -= root.val;
//     const ans = targetSum === 0 ? 1 : 0;
//     return ans + dfs(root.left, targetSum) + dfs(root.right, targetSum);
// }

// 前缀和解决
export function pathSum(root: TreeNode | null, targetSum: number): number {
    const map = new Map<number, number>();
    map.set(0, 1);

    const dfs = (node: TreeNode | null, curSum: number) => {
        if (!node) {
            return 0;
        }
        let ans = 0;
        curSum += node.val;
        ans += map.has(curSum - targetSum) ? map.get(curSum - targetSum) : 0;

        map.set(curSum, (map.has(curSum) ? map.get(curSum) : 0) + 1);

        const l = dfs(node.left, curSum);
        const r = dfs(node.right, curSum);

        ans += l + r;

        map.set(curSum, map.get(curSum) - 1);

        return ans;
    };

    return dfs(root, 0);
}

File: src/09.tree/05.437.go
package tree

// 双遍历
// func dfs1(root *TreeNode, targetSum int) int {
// 	if root == nil {
// 		return 0
// 	}

// 	targetSum -= root.Val

// 	ans := 0
// 	if targetSum == 0 {
// 		ans = 1
// 	}

// 	return ans + dfs1(root.Left, targetSum) + dfs1(root.Right, targetSum)
// }
// func PathSum(root *TreeNode, targetSum int) int {
// 	if root == nil {
// 		return 0
// 	}
// 	return dfs1(root, targetSum) + PathSum(root.Left, targetSum) + PathSum(root.Right, targetSum)
// }

// 前缀和
var m = map[int]int{}

func dfs1(node *TreeNode, curSum int, targetSum int) int {
	if node == nil {
		return 0
	}

	ans := 0
	curSum += node.Val
	if v, ok := m[curSum-targetSum]; ok {
		ans += v
	}

	m[curSum]++

	l := dfs1(node.Left, curSum, targetSum)
	r := dfs1(node.Right, curSum, targetSum)

	ans += l + r

	m[curSum]--

	return ans
}

func PathSum(root *TreeNode, targetSum int) int {
	m[0] = 1
	return dfs1(root, 0, targetSum)
}
